diff --git a/std_msgs/CMakeLists.txt b/std_msgs/CMakeLists.txt
index 354f1a9..1e14514 100644
--- a/std_msgs/CMakeLists.txt
+++ b/std_msgs/CMakeLists.txt
@@ -45,6 +45,7 @@ set(msg_files
   "msg/UInt64MultiArray.msg"
   "msg/UInt8.msg"
   "msg/UInt8MultiArray.msg"
+  "msg/Stamped1kb.msg"
 )
 rosidl_generate_interfaces(${PROJECT_NAME}
   ${msg_files}
diff --git a/std_msgs/msg/Stamped1kb.msg b/std_msgs/msg/Stamped1kb.msg
new file mode 100644
index 0000000..271bfab
--- /dev/null
+++ b/std_msgs/msg/Stamped1kb.msg
@@ -0,0 +1,5 @@
+builtin_interfaces/Time stamp
+uint32 tracking_number
+float32 frequency
+uint32 size
+byte[1024] data
