diff --git a/rcl/src/rcl/publisher.c b/rcl/src/rcl/publisher.c
index 62bf16b..67fc609 100644
--- a/rcl/src/rcl/publisher.c
+++ b/rcl/src/rcl/publisher.c
@@ -34,6 +34,8 @@ extern "C"
 #include "./common.h"
 #include "./publisher_impl.h"
 
+#include <ros2profiling/profiling.h>
+
 rcl_publisher_t
 rcl_get_zero_initialized_publisher()
 {
@@ -283,6 +285,7 @@ rcl_publish(
   const void * ros_message,
   rmw_publisher_allocation_t * allocation)
 {
+  store_profile(publisher->impl->rmw_handle->topic_name, (void*)ros_message, PROFIDX_PUB_RCL_PUBLISH, "rcl_publish", NULL);
   if (!rcl_publisher_is_valid(publisher)) {
     return RCL_RET_PUBLISHER_INVALID;  // error already set
   }
diff --git a/rcl/src/rcl/subscription.c b/rcl/src/rcl/subscription.c
index 2e3b37d..c314aed 100644
--- a/rcl/src/rcl/subscription.c
+++ b/rcl/src/rcl/subscription.c
@@ -32,6 +32,8 @@ extern "C"
 #include "./common.h"
 #include "./subscription_impl.h"
 
+#include <ros2profiling/profiling.h>  
+
 
 rcl_subscription_t
 rcl_get_zero_initialized_subscription()
@@ -254,6 +256,8 @@ rcl_take(
   rmw_subscription_allocation_t * allocation
 )
 {
+  uint64_t ts = get_timestamp(); 
+  
   RCUTILS_LOG_DEBUG_NAMED(ROS_PACKAGE_NAME, "Subscription taking message");
   if (!rcl_subscription_is_valid(subscription)) {
     return RCL_RET_SUBSCRIPTION_INVALID;  // error message already set
@@ -280,6 +284,8 @@ rcl_take(
   if (!taken) {
     return RCL_RET_SUBSCRIPTION_TAKE_FAILED;
   }
+  store_profile(subscription->impl->rmw_handle->topic_name, ros_message, PROFIDX_SUB_RCL_TAKE_ENTER, "rcl_enter", &ts);
+  store_profile(subscription->impl->rmw_handle->topic_name, ros_message, PROFIDX_SUB_RCL_TAKE_LEAVE, "rcl_leave", NULL);  
   return RCL_RET_OK;
 }
 
